library/core/src/array/mod.rs

```diff
@@ -197,6 +198,30 @@ impl<'a, T, const N: usize> IntoIterator for &'a mut [T; N] {
     }
 }
 
+#[stable(feature = "index_trait_on_arrays", since = "1.47.0")]
+impl<T, I, const N: usize> Index<I> for [T; N]
+where
+    [T]: Index<I>,
+{
+    type Output = <[T] as Index<I>>::Output;
+
+    #[inline]
+    fn index(&self, index: I) -> &Self::Output {
+        Index::index(self as &[T], index)
+    }
+}
+
+#[stable(feature = "index_trait_on_arrays", since = "1.47.0")]
+impl<T, I, const N: usize> IndexMut<I> for [T; N]
+where
+    [T]: IndexMut<I>,
+{
+    #[inline]
+    fn index_mut(&mut self, index: I) -> &mut Self::Output {
```

Comment:
Can we do comparison for `const N` to ensure that `I` is less than `N`?

Reply:
`I` is a type and `N` is a `const`, so they can't be compared directly. Also, it's not currently possible to use const generic parameters in const evaluation contexts. If at some point the aforementioned limitation to const generics is lifted and some kind of a bounded integer type or a trait is added to the standard library, then the standard library can also add a specialization for indexing arrays with that bounded type/trait which omits bounds checking.